pennylane/_qubit_device.py

```diff
@@ -1014,13 +1015,29 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
                 if op.name not in ("QubitStateVector", "BasisState", "Snapshot"):
                     expanded_ops.append(op)
 
-        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+        trainable_params = []
+        for k in tape.trainable_params:
+            # pylint: disable=protected-access
+            if hasattr(tape._par_info[k]["op"], "return_type"):
+                warnings.warn(
```

Comment:
Note: this warning will appear on the forward pass evaluation too if a user has:
* `diff_method="adjoint"` because `mode="forward"` is the default (hence gradients are computed under the hood);
* Didn't explicitly specify `requires_grad=False` for the matrix of Hermitian.

Having said that, the warning should be clear to encourage users to define `requires_grad=False`, so this behaviour should be acceptable.